### ISSUE TITLE
Use BinaryBuilder for MAGMA

### ISSUE THREAD

**Issue by Roger-luo** [Roger-luo/MAGMA.jl#4]
This is an example script made with BinaryBuilder, we could do this on bors with nvcc and provide downloads in github releases:

https://github.com/bicycle1885/ZlibBuilder

**Comment by JerryChen97** [comment]
When I was building MAGMA with BinaryBuilder, it seemed to have to find a right path for CUDA's lib. I think we may not be possible to provide MAGMA with a mere path without CUDA's installation in the sandbox.

Should we still build it with a CUDA as one of the dependencies? Or is there any other possible solution... @Roger-luo

**Comment by Roger-luo** [comment]
No, CUDA should be system default, if there isn't any CUDA installed in the PATH, then error. If there is then build.

**Comment by Roger-luo** [comment]
@staticfloat I heard you are working on BB for cublas etc. but what do you suggest we have here? just build magma from scratch?

**Comment by staticfloat** [comment]
Yes.  Actually, I got kind of distracted and ended up working on a [MAGMA builder](https://github.com/JuliaPackaging/Yggdrasil/blob/master/MAGMA/build_tarballs.jl) myself.  I managed to get most of the way through, but at the very end it was failing to link against `libopenblas64.so` properly.  This was because of our symbol renaming (we rename `dgemv_` to `dgemv_64_` to show that it is ILP64 capable) and so I [asked for help](https://icl.cs.utk.edu/magma/forum/viewtopic.php?f=2&t=3890&sid=bfbe5458d1f584567a3dc98199da38bb) on the MAGMA forums about doing symbol renaming, and I've only implemented half of their advice.  If you and @JerryChen97 want to pick this up and continue working with it, that would be great, as I ran out of time to continue working on this.

**Comment by JerryChen97** [comment]
@staticfloat Thank you so much! I will pick it up ASAP. I'll let you know once we continue with it.

**Comment by JerryChen97** [comment]
@staticfloat Hi Elliot~ Sorry for so late update on our project for MAGMA, but I have been struggling with lots of personal stuff and didn't have time to continue on the installation problem for months...
And I am wondering how can I get your build_tarballs.jl because it seems to be a 404 now...

**Comment by staticfloat** [comment]
Okay, so I looked into this again, and all that happened is that Yggdrasil had everything get re-named into directories based on first letter (so `MAGMA/build_tarballs.jl` is now located at `M/MAGMA/build_tarballs.jl`.)

The symbol renaming issue persists though, so I would ask you to first read [this forum link](https://icl.cs.utk.edu/magma/forum/viewtopic.php?f=2&t=3890&sid=bfbe5458d1f584567a3dc98199da38bb) and see if you understand the problem, then download the tarballs for [OpenBLAS](https://github.com/JuliaBinaryWrappers/OpenBLAS_jll.jl/releases/tag/OpenBLAS-v0.3.7%2B0) and [CUDA](https://github.com/JuliaBinaryWrappers/CUDA_jll.jl/releases/tag/CUDA-v10.1.243%2B0), unpack them together into some directory (like `~/magma_build/output`), then configure MAGMA to build on your native machine against those libraries.  OpenBLAS_jll has the same symbol renaming that Julia uses, so if you can figure out the proper patch to apply to MAGMA in order to get it to link against ILP64 OpenBLAS (where the symbols have `_64` appended to the end)then that would be very helpful.

**Comment by staticfloat** [comment]
So no need to mess with BinaryBuilder just yet; I just want you to build on your machine, locally.  You can use the same MAGMA [source tarball we use in the builder](http://icl.utk.edu/projectsfiles/magma/downloads/magma-2.5.1.tar.gz), but if you do that, make sure to run `git init && git add . && git commit -a -m "Initial commit"`) so that any changes you make to source files you can see with a simple `git diff -p`.

**Comment by JerryChen97** [comment]
@staticfloat Oh good, I will check the forum at first to see if I can really get the problem. Thank you so much!!!!

**Comment by JerryChen97** [comment]
@staticfloat I [tried the solution](https://github.com/Roger-luo/MAGMA/pull/1) in the forum discussion but the linker seems not able to locate these functions with suffix 64_.
I checked the error message to find that they mostly came from testing_*.cpp.
Here are some typical lines
`g++ -fPIC                       -fopenmp -Wl,-rpath,/home/jerry/TestsForMagma/MAGMA/lib \
-o testing/testing_zaxpy testing/testing_zaxpy.o \
-L./testing -ltest \
-L./lib -lmagma \
-L./testing/lin -llapacktest \
-L/usr/local/cuda/lib64 -L/opt/OpenBLAS/lib/lib -lopenblas -lcublas -lcusparse -lcudart -lcudadevrt
g++ -fPIC                       -fopenmp -Wl,-rpath,/home/jerry/TestsForMagma/MAGMA/lib \
-o testing/testing_zgemm testing/testing_zgemm.o \
-L./testing -ltest \
-L./lib -lmagma \
-L./testing/lin -llapacktest \
-L/usr/local/cuda/lib64 -L/opt/OpenBLAS/lib/lib -lopenblas -lcublas -lcusparse -lcudart -lcudadevrt
g++ -fPIC                       -fopenmp -Wl,-rpath,/home/jerry/TestsForMagma/MAGMA/lib \
-o testing/testing_zgemv testing/testing_zgemv.o \
-L./testing -ltest \
-L./lib -lmagma \
-L./testing/lin -llapacktest \
-L/usr/local/cuda/lib64 -L/opt/OpenBLAS/lib/lib -lopenblas -lcublas -lcusparse -lcudart -lcudadevrt
g++ -fPIC                       -fopenmp -Wl,-rpath,/home/jerry/TestsForMagma/MAGMA/lib \
-o testing/testing_zhemv testing/testing_zhemv.o \
-L./testing -ltest \
-L./lib -lmagma \
-L./testing/lin -llapacktest \
-L/usr/local/cuda/lib64 -L/opt/OpenBLAS/lib/lib -lopenblas -lcublas -lcusparse -lcudart -lcudadevrt
testing/testing_zaxpy.o: In function `main':
testing_zaxpy.cpp:(.text.startup+0xba): undefined reference to `dlamch64_'
testing_zaxpy.cpp:(.text.startup+0x2ce): undefined reference to `zlarnv64_'
testing_zaxpy.cpp:(.text.startup+0x2e3): undefined reference to `zlarnv64_'
testing_zaxpy.cpp:(.text.startup+0x336): undefined reference to `zlange64_'
testing_zaxpy.cpp:(.text.startup+0x361): undefined reference to `zlange64_'
testing_zaxpy.cpp:(.text.startup+0x580): undefined reference to `zaxpy64_'
testing_zaxpy.cpp:(.text.startup+0x5df): undefined reference to `zaxpy64_'
testing_zaxpy.cpp:(.text.startup+0x60a): undefined reference to `zlange64_'
`

**Comment by JerryChen97** [comment]
Not all testing cpps reported this sort of errors. All the other cpps only complained along with warnings. I am not sure what was going on.

**Comment by staticfloat** [comment]
You can see what the actual names are with `nm`:

```
$ nm libopenblas64_.so | grep zaxpy | head -2
0000000000115ca0 T cblas_zaxpy64_
0000000000115120 T zaxpy_64_
```

So it looks to me like we need another underscore _before_ the `64`.

**Comment by JerryChen97** [comment]
Corrected the naming problem, but now there is another error:
`testing/testing_zgeqp3_gpu.o: In function `main':
testing_zgeqp3_gpu.cpp:(.text.startup+0x774): undefined reference to `zqpt01_64_'
collect2: error: ld returned 1 exit status
Makefile:680: recipe for target 'testing/testing_zgeqp3_gpu' failed
`

**Comment by JerryChen97** [comment]
I think I almost solved the major bug in creating `libmagma.so` linked with `libopenblas64_.so`. It's a not-long but zig-zag way. 
Since it's too late and tomorrow morning I still have a course to take, tomorrow afternoon I will update it.

**Comment by JerryChen97** [comment]
I put the available mangling editing file on the make in our [MAGMA repo](https://github.com/Roger-luo/MAGMA/pull/3).

In short, we can just build the correct and usable MAGMA shared lib by commenting out the cmake mangling file, edit the ADD_ variable and in the make.inc file adding a series of correct flags to CFLAGS, CXXFLAGS, NVCCFLAGS, FFLAGS and F90FLAGS (the suggestion in MAGMA forum is a little bit faulty and there should not be any semicolon for the NVCCFLAGS options) and at last LIB=-lopenblas64_.

Meanwhile our MAGMA.jl requires a little bit editing before properly running but I found simply replacing every CInt & Int32 with Int solved the problem. I think maybe it is because the ilp64 lib does not support Int32 interface; in the past, I used lots of CInt or Int32 to conveniently avoid some converting bugs. Now after editing actually the MAGMA.jl wrapper will be more easy-to-read and consistent.

I think the next step might be to follow the [BB script](https://github.com/JuliaPackaging/Yggdrasil/blob/4b2fe25a051cd04c451a1677d0e2fa1834490dcd/M/MAGMA/build_tarballs.jl), is it okay? @staticfloat

**Comment by JerryChen97** [comment]
Just created a build_tarballs.jl in my own [repository](https://github.com/JerryChen97/MAGMABuilder/pull/1). However, my BinaryBuilder.run_wizard broke so currently I still have some problems not yet solved (e.g. the sha256 of src code) and not sure about whether the script works or not.

**Comment by JerryChen97** [comment]
The [recent CUDA builder](https://github.com/JuliaPackaging/Yggdrasil/blob/a1fd9a5cdec46e301d58f11699374686821b65e0/C/CUDA/CUDA_full%4010.2/build_tarballs.jl) seems to be built on the full version. I will try it ASAP.

**Comment by JerryChen97** [comment]
I [tried](https://github.com/JerryChen97/MAGMABuilder/blob/master/build_tarballs.jl) the `CUDA_full` as the dependency replacing the old `CUDA` and it seemed to be working quite good in the default Linux platform `Linux(:x86_64, libc=:glibc)`. However, all the other failed strangely. But considering the fact that this is the most common one I think we can just let it go.

**Comment by pnavaro** [comment]
Dear Jerry, is it possible to try on my side ?

Thanks

**Comment by JerryChen97** [comment]
> Dear Jerry, is it possible to try on my side ?
> 
> Thanks

Hi Pierre! I haven't tested that BB script yet, but I suppose it should work properly on any `x86_64` with `glibc` if we don't care about the great time consumption. However, the build for any other platforms failed so if you are using something else I am not sure whether or not it will work.

**Comment by pnavaro** [comment]
Sorry I did not read carefully the instructions given by Jerry. I tried on Ubuntu 18.04 
I get this error
~~~
./include/magma_types.h:63:14: fatal error: cuda.h: No such file or directory
     #include <cuda.h>    // for CUDA_VERSION
              ^~~~~~~~
compilation terminated.
make: *** [Makefile:609: control/magma_yield.o] Error 1
make: *** [Makefile:609: control/magma_zbulge.o] Error 1
Previous command exited with 2
ERROR: LoadError: Build for MAGMA on x86_64-linux-gnu did not complete successfully
~~~

**Comment by JerryChen97** [comment]
> Sorry I did not read carefully the instructions given by Jerry. I tried on Ubuntu 18.04
> I get this error
> 
> ```
> ./include/magma_types.h:63:14: fatal error: cuda.h: No such file or directory
>      #include <cuda.h>    // for CUDA_VERSION
>               ^~~~~~~~
> compilation terminated.
> make: *** [Makefile:609: control/magma_yield.o] Error 1
> make: *** [Makefile:609: control/magma_zbulge.o] Error 1
> Previous command exited with 2
> ERROR: LoadError: Build for MAGMA on x86_64-linux-gnu did not complete successfully
> ```

Oh this is definitely the same error I got in the very beginning for the other platforms. I will go check the script again to make it work.

**Comment by JerryChen97** [comment]
Hi Elliot @staticfloat , I am trying to work with the CUDA_full in [this script](https://github.com/JerryChen97/MAGMABuilder/blob/master/build_tarballs.jl), but it doesn't work and throws an error `nvcc not found`.
I suppose it is because `nvcc` is not yet provided by the `BinaryBuilder` sandbox, is it?
If so, is there any way for me to get around it?

**Comment by JerryChen97** [comment]
PS: the `cuda.h: No such file or directory` error is already solved and it was caused by my typo of the cuda path in the sandbox of BB...

**Comment by JerryChen97** [comment]
Hi Pierre @pnavaro , my current thought is that even though the `cuda.h: No such file or directory` bug is already fixed, the new problem `nvcc not found` should be quite non-trivial... I am so sorry about the previous overconfidence on the build script I've made but maybe you still have to wait for quite a time before we eventually fix everything.

**Comment by JerryChen97** [comment]
OK I found the nvcc in the CUDA/bin directory...

**Comment by JerryChen97** [comment]
I edited it and made it work without the previous bugs anymore. However, the BB script for all the other platforms still doesn't work due to the failure of `CUDA_full` for which I don't know why.

**Comment by pnavaro** [comment]
Thank you Jerry
The MAGMA library is built without error but I have a problem with the
shared library. 
```
/scratch/navaro/MAGMABuilder/build/x86_64-linux-gnu/TZcT23qV/destdir/lib/libmagma.so cannot be dlopen'ed
```
I tried to build MAGMA separately  using your recipe. It works well.
Thank you for working on this. No hurry.

**Comment by pnavaro** [comment]
fortran is not installed on the server il think this is the problem. Sorry for the noise.

**Comment by JerryChen97** [comment]
> Thank you Jerry
> The MAGMA library is built without error but I have a problem with the
> shared library.
> 
> ```
> /scratch/navaro/MAGMABuilder/build/x86_64-linux-gnu/TZcT23qV/destdir/lib/libmagma.so cannot be dlopen'ed
> ```
> 
> I tried to build MAGMA separately using your recipe. It works well.
> Thank you for working on this. No hurry.

Oh, actually this is indeed the problem I am currently trying to tackle. The script ran quite properly on my local machine but failed on the [Yggdrasil](https://github.com/JuliaPackaging/Yggdrasil/pull/874) along with a [quite similar error](https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=3112&view=logs&j=0d7d5c20-8585-5db2-71cf-4d4d2c46b238&t=3e28a03b-82d0-5971-66f0-5e5fbe595c00&l=7902).

Did you eventually find a solution to it?

**Comment by pnavaro** [comment]
Unfortunately not.

**Comment by JerryChen97** [comment]
> Unfortunately not.

Does it work on your own PC? I tried running it on my own Ubuntu 19.10 with CUDA 10.2 installed but nothing error happened; so currently it is very difficult to handle for me since I cannot even reproduce...

**Comment by pnavaro** [comment]
No it is not my PC, I have a Mac :-)  I have access to a server with a TESLA V100 and It is CUDA 10.0. I will ask if it is possible to upgrade cuda . But Ubuntu 18.04 can't be changed. I have also an access to another HPC server. I will try on Monday.

**Comment by pnavaro** [comment]
Do you need to be privileged user to execute the script ? because it asks me the sudo password...

**Comment by pnavaro** [comment]
I made the test on Fedora 29 with CUDA 10.2 and it goes to the end. I just have only one error
"cannot allocate memory in static TLS block" for libtsan.so.0.0.0. But It didn't stop and  generate the tar file.

**Comment by JerryChen97** [comment]
> Do you need to be privileged user to execute the script ? because it asks me the sudo password...

I don't remember I have ever granted that script a `sudo`... This is quite strange...

**Comment by JerryChen97** [comment]
> I made the test on Fedora 29 with CUDA 10.2 and it goes to the end. I just have only one error
> "cannot allocate memory in static TLS block" for libtsan.so.0.0.0. But It didn't stop and generate the tar file.

Could you share the full output log file? Maybe we can look into it together to see if there's any solution.

**Comment by pnavaro** [comment]
Here the log on the V100 server http://dl.free.fr/scpKW2kmh

**Comment by JerryChen97** [comment]
> Here the log on the V100 server http://dl.free.fr/scpKW2kmh

Thanks for your sharing. I will try to understand what happened there.

**Comment by JerryChen97** [comment]
> I made the test on Fedora 29 with CUDA 10.2 and it goes to the end. I just have only one error
> "cannot allocate memory in static TLS block" for libtsan.so.0.0.0. But It didn't stop and generate the tar file.

Since it finally generated the tar file, can you use it as usual in the [MAGMA.jl](https://github.com/Roger-luo/MAGMA.jl)?

**Comment by pnavaro** [comment]
Some tests od MAGMA.jl failed because it is not possible to compare CuArrays and Arrays with 
isapprox but the external magma library installed with the tar file works.

**Comment by JerryChen97** [comment]
> Some tests of MAGMA.jl failed because it is not possible to compare CuArrays and Arrays with
> isapprox but the external magma library installed with the tar file works.

Yeah I just got the same results.

I feel that the `isapprox` error is basically what you have mentioned before. I didn't try `MAGMA.jl` on CUDA10.2 until just now so had no idea before. I will try your previous solution to see if we can get rid of it.

Anyway, this is good news that we have a somehow available BB file, even though it still doesn't pass `Yggdrasil`'s build; that might take quite a long time for me to reach the final solution.
